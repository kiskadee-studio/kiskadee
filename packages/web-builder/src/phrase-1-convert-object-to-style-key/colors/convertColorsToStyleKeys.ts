import type {
  SemanticColor,
  InteractionState,
  ColorProperty,
  StyleKeyByElement,
  ElementStyle,
  PaletteName
} from '@kiskadee/schema';
import { update } from 'lodash';

/**
 * Generates a nested map of style key strings from a given palettes definition.
 *
 * Each palette contains color properties (e.g., background, text), each either:
 *   - an InteractionStateColorMap (with 'rest', 'hover', etc.)
 *   - a map of semantic colors (primary, secondary, danger, etc.), each mapping to an InteractionStateColorMap.
 *
 * A direct InteractionStateColorMap (identified by the presence of 'rest' at the top level)
 * is treated as a single "neutral" semantic color entry. This ensures that palettes without
 * explicit semantic variants still produce a consistent structure under the semanticColor key "neutral".
 *
 * For each color property, semantic color, and interaction state, a style key is generated
 * encoding the property, state, and color value.
 *
 * Style key format:
 *   - Rest state:
 *       "{property}__[<optional 'ref::'>]{JSON-stringified color}"
 *   - Other states:
 *       "{property}--{state}[<optional '::ref'>]__{JSON-stringified color}"
 *
 * A color reference (an object with a 'ref' field) is indicated by:
 *   - 'ref::' prefix on rest keys
 *   - '::ref' suffix on non-rest keys
 *
 * @param palettes  Object mapping palette names to color property definitions, each defining
 *                  either an InteractionStateColorMap or a SemanticColorMap.
 * @returns A nested map: paletteName → semanticColor → interactionState → array of style keys.
 */
export function convertColorsToStyleKeys(
  palettes: ElementStyle['palettes']
): StyleKeyByElement['palettes'] {
  const styleKeys: StyleKeyByElement['palettes'] = {};

  for (const p in palettes) {
    const paletteName = p as PaletteName;
    const colorSchema = palettes[paletteName];

    for (const c in colorSchema) {
      const colorProperty = c as ColorProperty;
      const colorEntry = colorSchema[colorProperty];
      if (colorEntry === undefined) continue;

      const isInteractionState = 'rest' in colorEntry;
      const semanticColorMap = isInteractionState ? { neutral: colorEntry } : colorEntry;

      for (const s in semanticColorMap) {
        const semanticColor = s as SemanticColor;
        const interactionStateMap = semanticColorMap[semanticColor];
        for (const i in interactionStateMap) {
          const interactionState = i as InteractionState;
          const colorValue = interactionStateMap[interactionState];
          const hasRef = typeof colorValue === 'object' && 'ref' in colorValue;
          const color = JSON.stringify(hasRef ? colorValue.ref : colorValue);

          const styleKey =
            interactionState === 'rest'
              ? `${colorProperty}__${hasRef ? 'ref::' : ''}${color}`
              : `${colorProperty}--${interactionState}${hasRef ? '::ref' : ''}__${color}`;

          update(
            styleKeys,
            [paletteName, semanticColor, interactionState],
            (arr: string[] = []) => {
              arr.push(styleKey);
              return arr;
            }
          );
        }
      }
    }
  }

  return styleKeys;
}
