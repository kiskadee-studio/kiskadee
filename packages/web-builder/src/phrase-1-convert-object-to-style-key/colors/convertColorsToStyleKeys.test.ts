import type { PaletteSchema } from '@kiskadee/schema';
import { describe, expect, it } from 'vitest';
import { convertColorsToStyleKeys } from './convertColorsToStyleKeys';

describe('convertColorsToStyleKeys', () => {
  it('generates style keys for palette property without reference', () => {
    const palettes: PaletteSchema = {
      p1: {
        boxColor: {
          primary: {
            rest: [45, 100, 50, 1]
          }
        }
      }
    };

    const result = convertColorsToStyleKeys(palettes);

    expect(result).toEqual({
      p1: {
        primary: {
          rest: ['boxColor__[45,100,50,1]']
        }
      }
    });
    expect(result).toMatchSnapshot();
  });

  it('generates style keys for palette property with a reference value', () => {
    const palettes: PaletteSchema = {
      p1: {
        borderColor: {
          primary: {
            rest: [255, 255, 255, 1],
            hover: { ref: [255, 255, 255, 0.1] }
          }
        }
      }
    };

    const result = convertColorsToStyleKeys(palettes);

    expect(result).toEqual({
      p1: {
        primary: {
          rest: ['borderColor__[255,255,255,1]'],
          hover: ['borderColor--hover::ref__[255,255,255,0.1]']
        }
      }
    });
    expect(result).toMatchSnapshot();
  });

  it('generates style keys for multiple palette entries', () => {
    const palettes: PaletteSchema = {
      p1: {
        textColor: {
          // propriedade > semantic > state
          primary: {
            rest: [120, 50, 50, 1],
            hover: { ref: [240, 50, 50, 0.5] }
          },
          secondary: {
            rest: [240, 50, 50, 0.5]
          }
        },
        borderColor: {
          primary: {
            rest: [120, 50, 50, 1]
          },
          danger: {
            rest: [0, 0, 0, 0.02],
            focus: { ref: [10, 20, 30, 0.1] }
          }
        }
      }
    };

    const result = convertColorsToStyleKeys(palettes);

    expect(result).toEqual({
      p1: {
        primary: {
          rest: ['textColor__[120,50,50,1]', 'borderColor__[120,50,50,1]'],
          hover: ['textColor--hover::ref__[240,50,50,0.5]']
        },
        secondary: {
          rest: ['textColor__[240,50,50,0.5]']
        },
        danger: {
          rest: ['borderColor__[0,0,0,0.02]'],
          focus: ['borderColor--focus::ref__[10,20,30,0.1]']
        }
      }
    });
    expect(result).toMatchSnapshot();
  });

  it('treats direct interaction-state map as neutral semantic color', () => {
    const palettes: PaletteSchema = {
      p1: {
        // Direct interaction‚Äêstate map: no semantic keys, should map to "neutral"
        boxColor: {
          rest: [0, 128, 255, 1],
          hover: { ref: [0, 128, 255, 0.5] }
        }
      }
    };

    const result = convertColorsToStyleKeys(palettes);

    expect(result).toEqual({
      p1: {
        neutral: {
          rest: ['boxColor__[0,128,255,1]'],
          hover: ['boxColor--hover::ref__[0,128,255,0.5]']
        }
      }
    });
    expect(result).toMatchSnapshot();
  });
});
