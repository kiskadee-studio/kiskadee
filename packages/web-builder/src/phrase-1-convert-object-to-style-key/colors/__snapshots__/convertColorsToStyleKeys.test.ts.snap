// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertColorsToStyleKeys > should process a palette property with a ref value 1`] = `
{
  "p1": {
    "primary": {
      "hover": [
        "borderColor--hover::ref__[255,255,255,0.1]",
      ],
      "rest": [
        "borderColor__[255,255,255,1]",
      ],
    },
  },
}
`;

exports[`convertColorsToStyleKeys > should process a palette property without ref 1`] = `
{
  "p1": {
    "primary": {
      "rest": [
        "boxColor__[45,100,50,1]",
      ],
    },
  },
}
`;

exports[`convertColorsToStyleKeys > should process multiple palette entries 1`] = `
{
  "p1": {
    "danger": {
      "focus": [
        "borderColor--focus::ref__[10,20,30,0.1]",
      ],
      "rest": [
        "borderColor__[0,0,0,0.02]",
      ],
    },
    "primary": {
      "hover": [
        "textColor--hover::ref__[240,50,50,0.5]",
      ],
      "rest": [
        "textColor__[120,50,50,1]",
        "borderColor__[120,50,50,1]",
      ],
    },
    "secondary": {
      "rest": [
        "textColor__[240,50,50,0.5]",
      ],
    },
  },
}
`;

exports[`convertColorsToStyleKeys > should treat a direct interaction‚Äêstate map as neutral semantic color 1`] = `
{
  "p1": {
    "neutral": {
      "hover": [
        "boxColor--hover::ref__[0,128,255,0.5]",
      ],
      "rest": [
        "boxColor__[0,128,255,1]",
      ],
    },
  },
}
`;
